File: clientC/src/pretest.c
#include "pretest.h"
#include "connection.h"
#include "tests.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define PRETEST_DURATION_NS 2000000000ULL   /* 2 s */
#define MIN_CHUNK           1024
#define MAX_CHUNK           (4 * 1024 * 1024)

static uint64_t now_ns(void)
{
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (uint64_t)ts.tv_sec * 1000000000ULL + ts.tv_nsec;
}

/* Thread-count tables: {min_mbps, threads}, descending. */
static const double DL_THRESH[] = { 100.0, 1.0, 0.0 };
static const int    DL_COUNT[]  = {   5,   3,   1   };
static const double UL_THRESH[] = { 150.0, 80.0, 30.0, 0.0 };
static const int    UL_COUNT[]  = {   5,    3,    2,   1   };

static int threads_for(double mbps, const double *thresh, const int *count, int n)
{
    for (int i = 0; i < n; i++)
        if (mbps >= thresh[i]) return count[i];
    return 1;
}

int run_pretest(const char *addr, uint16_t port,
                int use_tls, int no_tls_verify, int protocol,
                const char *token, int max_threads,
                PretestResult *out)
{
    printf("\nPre-test: measuring baseline throughput...\n");

    RmbtConn *conn = conn_connect(addr, port, use_tls, no_tls_verify, protocol);
    if (!conn) return -1;
    if (conn_greeting(conn, token) < 0) { conn_free(conn); return -1; }

    size_t server_min = conn->chunk_size_min;
    size_t server_max = conn->chunk_size_max;

    uint64_t t_start    = now_ns();
    size_t   cs         = server_min > MIN_CHUNK ? server_min : MIN_CHUNK;
    int      n          = 1;
    uint64_t last_bytes = 0;
    uint64_t last_ns    = 0;  /* TIME from the last (largest) GETCHUNKS batch */
    uint64_t rtt_ns     = 0;  /* TIME of the first (tiny) batch ≈ round-trip time */

    char line[256];

    for (;;) {
        if (now_ns() - t_start >= PRETEST_DURATION_NS) break;

        if (conn_read_line(conn, line, sizeof(line)) < 0) break;
        if (!strstr(line, "GETCHUNKS")) {
            fprintf(stderr, "pre-test: expected ACCEPT with GETCHUNKS, got: %s\n", line);
            conn_free(conn);
            return -1;
        }

        char cmd[64];
        snprintf(cmd, sizeof(cmd), "GETCHUNKS %d %zu", n, cs);
        if (conn_write_line(conn, cmd) < 0) break;

        unsigned char *buf = malloc(cs);
        if (!buf) break;
        for (int i = 0; i < n; i++) {
            if (conn_read_exact(conn, buf, cs) < 0) {
                free(buf);
                goto done;
            }
        }
        free(buf);
        last_bytes = (uint64_t)n * cs;

        if (conn_write_line(conn, "OK") < 0) break;
        if (conn_read_line(conn, line, sizeof(line)) < 0) break;
        {
            uint64_t t;
            if (parse_time_ns(line, &t) == 0) {
                if (rtt_ns == 0) rtt_ns = t;  /* first tiny batch ≈ RTT */
                last_ns = t;
            }
        }

        /* Exponential progression */
        if (n >= 8) {
            cs = cs * 2;
            size_t cap = server_max < MAX_CHUNK ? server_max : MAX_CHUNK;
            if (cs > cap) cs = cap;
            n = 1;
        } else {
            n *= 2;
        }
    }

done:;
    /* TIME = transmission_time + RTT.  Subtract the RTT estimate (first tiny
     * batch, where transmission_time ≈ 0) to get actual throughput. */
    uint64_t transfer_ns = (last_ns > rtt_ns) ? last_ns - rtt_ns : 0;
    double bps;
    if (transfer_ns > 0)
        bps = last_bytes / (transfer_ns / 1e9);
    else if (last_ns > 0)
        bps = last_bytes / (last_ns / 1e9);
    else
        bps = last_bytes / ((now_ns() - t_start) / 1e9);
    double mbps = bps * 8.0 / 1e6;

    /* Target 50 chunks/sec → round to nearest KiB, clamp to [min, max]. */
    size_t ideal   = (size_t)(bps / 50.0);
    size_t rounded = ((ideal + 512) / 1024) * 1024;
    size_t cap     = server_max < MAX_CHUNK ? server_max : MAX_CHUNK;
    size_t floor_v = server_min > MIN_CHUNK ? server_min : MIN_CHUNK;
    size_t chunk_size = rounded < floor_v ? floor_v :
                        rounded > cap     ? cap      : rounded;

    int dl = threads_for(mbps, DL_THRESH, DL_COUNT, 3);
    int ul = threads_for(mbps, UL_THRESH, UL_COUNT, 4);
    if (dl > max_threads) dl = max_threads;
    if (ul > max_threads) ul = max_threads;

    printf("  pre-test: %.1f Mbit/s → chunk=%zu KiB  dl_threads=%d  ul_threads=%d\n",
           mbps, chunk_size / 1024, dl, ul);

    conn_quit(conn);
    conn_free(conn);

    out->chunk_size = chunk_size;
    out->dl_threads = dl;
    out->ul_threads = ul;
    return 0;
}

File: src/pretest.rs
use anyhow::{bail, Result};
use std::time::{Duration, Instant};

use crate::connection::{RmbtConn, Protocol};

const PRETEST_DURATION: Duration = Duration::from_millis(2000);
const MIN_CHUNK: usize = 1024;
const MAX_CHUNK: usize = 4 * 1024 * 1024;

pub struct PretestResult {
    pub chunk_size: usize,
    pub dl_threads: usize,
    pub ul_threads: usize,
}

// Rows are (min_mbps, thread_count), sorted descending so the first match wins.
const DL_TABLE: &[(f64, usize)] = &[(100.0, 5), (1.0, 3), (0.0, 1)];
const UL_TABLE: &[(f64, usize)] = &[(150.0, 5), (80.0, 3), (30.0, 2), (0.0, 1)];

fn threads_for(mbps: f64, table: &[(f64, usize)]) -> usize {
    table.iter()
         .find(|&&(thresh, _)| mbps >= thresh)
         .map(|&(_, c)| c)
         .unwrap_or(1)
}

/// Run a 2-second single-thread download pre-test using GETCHUNKS with
/// exponential chunk-count and chunk-size progression.
/// Returns the optimal chunk size and suggested thread counts for each direction.
pub fn run_pretest(
    addr:          &str,
    port:          u16,
    use_tls:       bool,
    no_tls_verify: bool,
    protocol:      Protocol,
    token:         &str,
    max_threads:   usize,
) -> Result<PretestResult> {
    println!("\nPre-test: measuring baseline throughput…");

    let mut conn = RmbtConn::connect(addr, port, use_tls, no_tls_verify, protocol)?;
    conn.greeting(token)?;

    let server_min = conn.chunk_size_min;
    let server_max = conn.chunk_size_max;

    let t_start   = Instant::now();
    let mut cs    = server_min.max(MIN_CHUNK);
    let mut n     = 1usize;
    let mut total = 0u64;

    loop {
        if t_start.elapsed() >= PRETEST_DURATION { break; }

        let accept = conn.read_accept()?;
        if !accept.contains("GETCHUNKS") {
            bail!("pre-test: expected ACCEPT with GETCHUNKS, got: {accept}");
        }

        conn.write_line(&format!("GETCHUNKS {n} {cs}"))?;

        let mut buf = vec![0u8; cs];
        for _ in 0..n {
            conn.read_exact(&mut buf)?;
        }
        total += (n as u64) * (cs as u64);

        conn.write_line("OK")?;
        let _ = conn.read_line()?; // discard TIME <ns>

        // Exponential progression: double n until n≥8, then double chunk size.
        if n >= 8 {
            cs = (cs * 2).min(server_max.min(MAX_CHUNK));
            n = 1;
        } else {
            n *= 2;
        }
    }

    let elapsed = t_start.elapsed();
    let bps     = total as f64 / elapsed.as_secs_f64();
    let mbps    = bps * 8.0 / 1_000_000.0;

    // Target 50 chunks/sec (1 chunk per 20 ms), rounded to the nearest KiB.
    let ideal     = (bps / 50.0) as usize;
    let rounded   = ((ideal + 512) / 1024) * 1024;
    let chunk_size = rounded
        .max(server_min.max(MIN_CHUNK))
        .min(server_max.min(MAX_CHUNK));

    let dl = threads_for(mbps, DL_TABLE).min(max_threads);
    let ul = threads_for(mbps, UL_TABLE).min(max_threads);

    println!(
        "  pre-test: {mbps:.1} Mbit/s → chunk={} KiB  dl_threads={dl}  ul_threads={ul}",
        chunk_size / 1024
    );

    conn.quit()?;
    Ok(PretestResult { chunk_size, dl_threads: dl, ul_threads: ul })
}
